### ISSUE TITLE
Support static compilation

### ISSUE THREAD

**Issue by yihtserns** [yihtserns/groovy-decorator#3]
Currently when static compilation is enabled (e.g. via `@StaticCompile`), an exception will be thrown:

``` groovy
General error during class generation: size==0

java.lang.ArrayIndexOutOfBoundsException: size==0
    at org.codehaus.groovy.classgen.asm.OperandStack.getTopOperand(OperandStack.java:729)
    at org.codehaus.groovy.classgen.asm.BinaryExpressionHelper.evaluateEqual(BinaryExpressionHelper.java:306)
...
```

**Comment by yihtserns** [comment]
Don't know when, but I think I somehow fixed that.  But now I'm getting:

``` groovy
java.lang.ArrayIndexOutOfBoundsException: Internal compiler error while compiling script14574537688571698147625.groovy
Method: MethodNode@1147236197[java.lang.Object doCall(java.lang.Object)]
Line -1, expecting casting to java.lang.Object but operand stack is empty
    at org.codehaus.groovy.classgen.asm.OperandStack.doConvertAndCast(OperandStack.java:323)
...
```

**Comment by yihtserns** [comment]
Getting over that using 

``` groovy
java.lang.ArrayIndexOutOfBoundsException: Internal compiler error while compiling script14574534323081698147625.groovy
Method: MethodNode@1470373550[java.lang.Object doCall(java.lang.Object)]
Line -1, expecting casting to java.lang.Object but operand stack is empty
    at org.codehaus.groovy.classgen.asm.OperandStack.throwExceptionForNoStackElement(OperandStack.java:310)
    at org.codehaus.groovy.classgen.asm.OperandStack.doConvertAndCast(OperandStack.java:315)
...
```

...which wouldn't happen when using Groovy version 2.3.9.

**Comment by yihtserns** [comment]
Changed method interception strategy to use the one `groovy.transform.Memoized` uses: 2b454702f0b4c30a167fa7b818563c455046db0a.

Remaining:
- [x] Handle array(s) type parameter
- [x] Handle method with similar named parameter (e.g. Same class name, but different package)

**Comment by yihtserns** [comment]
Fixed with 172cd971e680b58eb03d2d30d67bb7d132508a5c, 5fbb717bd5666a5a2ad715274e9e6914ba4538ce, ee52524ecf2a2714dec35f23592b6a35e2311f2a, and 2b454702f0b4c30a167fa7b818563c455046db0a.
